engine/src/cmd/unit_json_factory.cpp

```diff
@@ -44,24 +40,45 @@ void UnitJSONFactory::ParseJSON(VSFileSystem::VSFile &file, bool player_ship) {
     } catch (std::exception const& e) {
         VS_LOG_FLUSH_EXIT(fatal, (boost::format("Error parsing JSON in UnitJSONFactory::ParseJSON(): %1%") % e.what()), 42);
     }
-    boost::json::array root_array = json_value.get_array();
-
-    for(boost::json::value& unit_value : root_array) {
-        boost::json::object unit_object = unit_value.get_object();
-        std::map<std::string, std::string> unit_attributes;
+    if (json_value.is_object()) {
+        boost::json::object obj = json_value.as_object();
+        std::map<std::string, std::string> unit_attributes{};
 
-        for(boost::json::key_value_pair& pair : unit_object) {
+        for (boost::json::key_value_pair& pair : obj) {
             const std::string value = boost::json::value_to<std::string>(pair.value());
             unit_attributes[pair.key()] = value;
         }
 
         // Add root
-        unit_attributes["root"] = file.GetRoot();  
+        unit_attributes["root"] = file.GetRoot();
 
-        if(player_ship) {
+        if (player_ship) {
             UnitCSVFactory::units["player_ship"] = unit_attributes;
         } else {
             UnitCSVFactory::units[unit_attributes["Key"]] = unit_attributes;
         }
+    } else if (json_value.is_array()) {
+        boost::json::array json_root = json_value.as_array();
+        for (boost::json::value & unit_value : json_root) {
+            boost::json::object unit_object = unit_value.get_object();
+            std::map<std::string, std::string> unit_attributes{};
+
+            for (boost::json::key_value_pair& pair : unit_object) {
+                const std::string value = boost::json::value_to<std::string>(pair.value());
+                unit_attributes[pair.key()] = value;
+            }
+
+            // Add root
+            unit_attributes["root"] = file.GetRoot();
+
+            if (player_ship) {
+                UnitCSVFactory::units["player_ship"] = unit_attributes;
+            } else {
+                UnitCSVFactory::units[unit_attributes["Key"]] = unit_attributes;
+            }
+        }
+    } else {
+        VS_LOG_FLUSH_EXIT(fatal, (boost::format("Saved game unit file '%1%' had an unexpected JSON structure. We don't know how to process it.") % file.GetFilename()), 42);
```

Comment:
I'm pretty sure this code also loads the ships from unit.json but not ships.json. That one is handled by UnitOptimizeFactory.

Reply:
Based on my testing, this was the loading code that needed to change.

Reply:
I'm not arguing that this code didn't need to change. I'm saying it could try and load `unit.json` and fail and tell the player "Saved game unit file..." for an asset file.

Reply:
Oh, so you're saying that the error message could be erroneous? Referring to a "saved game unit file" when that might not be what it is?

Reply:
Yes. It's a unit file but it could be either units.json or <save_game>.json.

Reply:
Corrected the error message.